### DISCUSSION TITLE
Using Coroutines in the BDD framework

### DISCUSSION THREAD

**Discussion by ekodl** [sbergen/Responsible#222] [Q&A]
Salutations,

We're attempting to use the BDD part of the Responsible framework to try to automate acceptance testing in our Unity application. We're running into an issue however, in that we're not sure how to use coroutines or async methods when executing things as part of the Given/When instructions. Is there a part of the documentation that we're missing? Or do you have any suggestions for usage? Is this even possible at this time?

For example, say I have the following coroutines where setup is occuring.

```csharp
public IEnumerator MySampleGivenCoroutine()
{
    yield return WaitForFrames(60);
}

public IEnumerator MySampleWhenCoroutine()
{
    yield return WaitForSeconds(2);
    myBool = true;
}
```

and that I wanted to test them as such:

```csharp
[UnityTest]
public IEnumerator MySampleBDDTest()
{
    yield return executor.YieldScenario(
        Scenario("I can do BDD tests with coroutines!"),
        Given("That I do a long running task", DoAndReturn<IEnumerator>("...", MySampleGivenCoroutine)),
        When("I wait two seconds", DoAndReturn<IEnumerator>("...", MySampleWhenCoroutine)),
        Then("myBool will be true", Do("...", () => Assert.That(myBool, Is.True)))
        );
}
```

However, when we execute this, it doesn't appear to actually run the coroutines, as it references 0 frames despite putting check marks next to the Given and When sections:
```
Failure context:
[!] Scenario: I can do BDD tests with coroutines! (Failed after 0.01 s ≈ 0 frames)
  [✓] Given That I do a long running task (Completed in 0.00 s ≈ 0 frames)
    [✓] ... (Completed in 0.00 s ≈ 0 frames)
  [✓] When I wait two seconds (Completed in 0.00 s ≈ 0 frames)
    [✓] ... (Completed in 0.00 s ≈ 0 frames)
  [!] Then myBool will be true (Failed after 0.01 s ≈ 0 frames)
    [!] ... (Failed after 0.01 s ≈ 0 frames)
 
      Failed with:
        NUnit.Framework.AssertionException:
            Expected: True
            But was:  False
```

Any ideas? Thank you for your assistance, we greatly appreciate your product.

**Comment by sbergen** [comment]
Hello,

The way to run a coroutine would be to use [Responsibly.WaitForCoroutine](https://www.beatwaves.net/Responsible/api/Responsible.Responsibly.html#Responsible_Responsibly_WaitForCoroutine_System_String_System_Func_System_Collections_IEnumerator__System_String_System_String_System_Int32_), together with expecting it to finish within some deadline with [ExpectWithinSeconds](https://www.beatwaves.net/Responsible/api/Responsible.TestWaitCondition.html#Responsible_TestWaitCondition_ExpectWithinSeconds__1_Responsible_ITestWaitCondition___0__System_Double_System_String_System_String_System_Int32_).

For async/await code, you can use [WaitForTask](https://www.beatwaves.net/Responsible/api/Responsible.Responsibly.html#Responsible_Responsibly_WaitForTask__1_System_String_System_Func_System_Threading_CancellationToken_System_Threading_Tasks_Task___0___System_String_System_String_System_Int32_).

It might feel a bit unintuitive to use wait conditions for this at first, but it is require to enforce a timeout on all operations, and provides better composability in the big picture.

And sorry for the late response! It seems like I need to tweak my notification settings to get notified on new discussions 🤔
